### PR TITLE
Support more system-site packages

### PR DESCRIPTION
**Pull request by orlitzky**
Add boilerplate for a few more python packages to be detected with `--enable-system-site-packages`.

Marked as critical because it repairs rpy2 system package on ubuntu/debian

### CODE REVIEW

**Comment by dimpase** [comment]
On Debian for `poetry_core` to work, one needs to install `python3-pep517`. Otherwise e.g. building `pkgconf` fails, as `import poetry` doesn't work.

**Comment by orlitzky** [comment]
> On Debian for `poetry_core` to work, one needs to install `python3-pep517`. Otherwise e.g. building `pkgconfig` fails, as `import poetry` doesn't work.

Hm, for now I've dropped the two `poetry_core` commits. Does that fix the issue?

Some time in the future I can refactor the existing python package check macro into something like `SAGE_RUN_PYTHON_IFELSE`, which would let us try `import poetry`, but the thought of doing it right now is giving me a headache.

**Comment by dimpase** [comment]
What I meant to say that the way you test for poetry is a bit buggy - with PEP517 packages one needs this extra package on Debian, otherwise the import just does not work.

although maybe it is a Debian bug.

**Comment by dimpase** [comment]
I reported this to Debian: https://bugs.debian.org/cgi-bin/bugreport.cgi?bug=1051833

**Comment by orlitzky** [comment]
> What I meant to say that the way you test for poetry is a bit buggy - with PEP517 packages one needs this extra package on Debian, otherwise the import just does not work.

I understood. What we'd like to do is run `import poertry` in addition to the usual test, to see if it works.

The existing python package check macro is a ton of boilerplate wrapped around a single command,

```m4
      AS_IF(
        [PYTHONUSERBASE="${PYTHONUSERBASE}" config.venv/bin/python3 -c dnl
           "import pkg_resources;                                      dnl
            pkg_resources.require('${SAGE_PKG_VERSPEC}'.splitlines())" dnl
	 2>&AS_MESSAGE_LOG_FD],
        [AC_MSG_RESULT(yes)],
        [AC_MSG_RESULT(no); sage_spkg_install_$1=yes]
      )
```

To run `import poetry`, we'd just have to replace the command. But obviously I don't want to copy/paste all of the boilerplate. Instead the macro could be refactored to take the command as an argument, but that gets a bit messy because of the interaction with the venv and such.

**Comment by mkoeppe** [comment]
The `rpy2` package also needs work -- adding the distro information (in #36141) causes the system package for rpy2 to be installed (it should not, in the standard configuration.)  
See https://github.com/sagemath/sage/actions/runs/6139868737: #36141 broke the CI runs for `ubuntu-{focal,jammy}`, `debian-{buster,bookworm,trixie,sid}`:  ("Unable to locate package rpy2")

**Comment by orlitzky** [comment]
> The `rpy2` package also needs work

fixed with prejudice

**Comment by mkoeppe** [comment]
That certainly does the job; but the real bug is that the package rpy2 is even selected in the standard configuration.
I think this would go away if you extended the rpy2 spkg-configure.m4 to use your macro

**Comment by orlitzky** [comment]
> That certainly does the job; but the real bug is that the package rpy2 is even selected in the standard configuration. I think this would go away if you extended the rpy2 spkg-configure.m4 to use your macro

Where is that package list coming from again?

I can replace the `spkg-configure.m4` if you think it will fix it. We don't have rpy2 in Gentoo yet though so I've never tested it.

**Comment by orlitzky** [comment]
> Where is that package list coming from again?

Oh, right, the grep. I've updated the `spkg-configure.m4` but haven't tested it. We don't even have rpy2 in the sage-on-gentoo overlay.

**Comment by mkoeppe** [comment]
Can't just delete the existing logic in rpy2's spkg-configure

**Comment by orlitzky** [comment]
> Can't just delete the existing logic in rpy2's spkg-configure

Again, I haven't tested it, but I didn't just delete that bit blindly. The stuff that was in there made sure that we used rpy2 from the system only when R itself came from the system. That used to be necessary because of the system-site-packages restriction. Now, rpy2 will only be detected when system site packages are enabled. And in that case the SPKG for R should be fine too.

**Comment by mkoeppe** [comment]
No, we have never used rpy2 from the system. The logic there is to build rpy2 when system R is found, and otherwise skip it. See #35347

**Comment by orlitzky** [comment]
> No, we have never used rpy2 from the system. The logic there is to build rpy2 when system R is found, and otherwise skip it. See #35347

Ok, trying again.

File: build/pkgs/rpy2/spkg-configure.m4 [hunk 1]
```diff
@@ -1,6 +1,6 @@
 SAGE_SPKG_CONFIGURE([rpy2], [
-    sage_spkg_install_rpy2=yes
-  ], [dnl REQUIRED-CHECK
+    SAGE_SPKG_CONFIGURE([rpy2], [SAGE_PYTHON_PACKAGE_CHECK([rpy2])])
```
Comment (mkoeppe) [inline build/pkgs/rpy2/spkg-configure.m4]: Nesting SAGE_SPKG_CONFIGURE within SAGE_SPKG_CONFIGURE? That can't be right

Comment (orlitzky) [reply build/pkgs/rpy2/spkg-configure.m4]: Ha, no. I was supposed to delete half of that after copy/pasting it in.

**Comment by mkoeppe** [comment]
Thanks, this works.
I think you can revert f85cc8a now

**Comment by orlitzky** [comment]
> Thanks, this works. I think you can revert [f85cc8a](https://github.com/sagemath/sage/commit/f85cc8a2c77c92cd0caea3b2cd319d573e99734d) now

Rebased away.

**Review by mkoeppe — APPROVED** [review]

**Comment by mkoeppe** [comment]
The struggles with `poetry` will go away after updates in #36129 and switching `pkgconfig` to a wheel package (#36181)

**Review by dimpase — APPROVED** [review]

**Comment by tobiasdiez** [comment]
On Conda the Conda provided rpy2 package seems to be rejected, see ci. Is this according to design?

Also there are a lot of errors that certain system packages couldn't be used.

**Comment by orlitzky** [comment]
> On Conda the Conda provided rpy2 package seems to be rejected, see ci. Is this according to design?
> 
> Also there are a lot of errors that certain system packages couldn't be used.

Nothing is being detected because `--enable-system-site-packages` was not passed to `./configure`. Do those conda CI jobs try to force all system packages? If so, they should probably use the sage package-listing utilities to avoid listing python packages when `--enable-system-site-packages` was not given.

More generally though, we just enabled hundreds of system packages without much consideration for whether or not the popular versions of those packages actually work for sage. Conda (or any distro) won't be expected to work until someone tries to install all of these packages as system packages and then runs through the test suite and fixes all of the bugs they hit. It's still an experimental feature that, more than anything, lets us start fixing said bugs without having to jump through hoops.

**Comment by dimpase** [comment]
I experimented with using this facility under Python 3.11.4 installed by pyenv.
I'm hitting a snag in form of which seem to tell something 
along the lines of https://stackoverflow.com/questions/63958574/no-template-sub-directory-with-name-lab-found-in-the-following-paths

Here are more `spkg-configure.m4` added:
[0001-more-spkg-configure.m4-for-python-packages.patch.txt](https://github.com/sagemath/sage/files/12623865/0001-more-spkg-configure.m4-for-python-packages.patch.txt)

```
[sagemath_doc_html-none] sage --docbuild --no-pdf-links reference/combinat inventory --no-prune-empty-dirs
[sagemath_doc_html-none] [repl     ] Exception occurred:
[sagemath_doc_html-none] [repl     ]   File "/home/dimpase/.pyenv/versions/3.11.5/lib/python3.11/site-packages/nbconvert/exporters/templateexporter.py", line 653, in get_template_names
[sagemath_doc_html-none] [repl     ]     raise ValueError(msg)
[sagemath_doc_html-none] [repl     ] ValueError: No template sub-directory with name 'script' found in the following paths:
[sagemath_doc_html-none] [repl     ]    /doesnotexist
[sagemath_doc_html-none] [repl     ]    /home/dimpase/work/software/sage/local/var/lib/sage/venv-python3.11/share/jupyter
[sagemath_doc_html-none] [repl     ]    /doesnotexist
[sagemath_doc_html-none] [repl     ]    /home/dima/.pyenv/share/jupyter
[sagemath_doc_html-none] [repl     ]    /usr/local/share/jupyter
[sagemath_doc_html-none] [repl     ]    /usr/share/jupyter
[sagemath_doc_html-none] [repl     ] The full traceback has been saved in /tmp/sphinx-err-mqe4c7d3.log, if you want to report the issue to the developers.
[sagemath_doc_html-none] [repl     ] Please also report this if it was a user error, so that a better error message can be provided next time.
[sagemath_doc_html-none] [repl     ] A bug report can be filed in the tracker at <https://github.com/sphinx-doc/sphinx/issues>. Thanks!
[sagemath_doc_html-none] Error building the documentation.
```

**Comment by orlitzky** [comment]
>     * not sure how it shoud be properly fixed. It seems to be a bug on the Jupyter/Python border.

I can't really say. I'm not a notebook user anyway, but I also haven't even started to sort out the notebook issues with system site packages enabled. If I had to bet though, I'd go with: version mismatches. Our ipython and notebook are outdated,  but many of their dependencies are missing bounds in `install-requires.txt`, so it's possible to wind up with an  `nbconvert` much newer than the notebook itself. Regardless, I'm thinking it's probably a waste of time until the notebook-7.x upgrade takes place.
 
> Here are more `spkg-configure.m4` added: [0001-more-spkg-configure.m4-for-python-packages.patch.txt](https://github.com/sagemath/sage/files/12623865/0001-more-spkg-configure.m4-for-python-packages.patch.txt)

I'm happy to have them but can we do it in a new ticket that isn't positively reviewed already? Note also that jupyter_packaging is being removed in https://github.com/sagemath/sage/pull/36267

**Comment by mkoeppe** [comment]
Follow-up ticket please...

**Comment by dimpase** [comment]
> Follow-up ticket please...

Please see #36276

**Pull request merged**